common/unsafe/src/main/java/org/apache/spark/unsafe/array/ByteArrayMethods.java

```diff
@@ -40,24 +41,41 @@ public static int roundNumberOfBytesToNearestWord(int numBytes) {
     }
   }
 
+  private static final boolean alignedArch = UnsafeAlignedOffset.getAlignedArch();
   /**
    * Optimized byte array equality check for byte arrays.
    * @return true if the arrays are equal, false otherwise
    */
   public static boolean arrayEquals(
       Object leftBase, long leftOffset, Object rightBase, long rightOffset, final long length) {
     int i = 0;
-    while (i <= length - 8) {
-      if (Platform.getLong(leftBase, leftOffset + i) !=
-        Platform.getLong(rightBase, rightOffset + i)) {
-        return false;
+
+    // check if stars align and we can get both offsets to be aligned
+    if ((leftOffset % 8 != 0) && ((leftOffset % 8) == (rightOffset % 8))) {
```

Comment:
I think this can actually be slightly simplified, to squeeze out one more branch, because the first condition ends up being redundant with what the while loop evaluates first anyway.

```
if ((leftOffset % 8) == (rightOffset % 8)) {
  while ((leftOffset + i) % 8 != 0 && i < length) {
```